src/motion_input_shaping/zero_vibration_stream_generator.py

```diff
@@ -0,0 +1,361 @@
+    # superposition of the contribution from each impulse and is equivalent to the convolution
+    shaped_acceleration = np.cumsum(accel_changes)  # Acceleration
+
+    return [shaped_time.tolist(), shaped_acceleration.tolist()]
+
+
+def create_stream_trajectory(
+    trajectory_time: list[float], trajectory_acceleration: list[float]
+) -> list[StreamSegment]:
+    """
+    Compute information needed to execute trajectory through streams.
+
+    Returns list of StreamSegment objects.
+    The final acceleration must be 0.
+
+    :param trajectory_time: Start time of each constant acceleration segment
+    :param trajectory_acceleration: Acceleration of each segment
+    """
+    stream_segments = []
+    # trajectory is one row less than list of accelerations since first row would be the
+    # initial position (zeros)
+    previous_position = 0.0
+    previous_velocity = 0.0
+    for n in range(0, len(trajectory_acceleration) - 1):
+        # Calculate position and velocity at end of each segment using equations for constant
+        # acceleration since acceleration changes are steps.
+        current_accel = trajectory_acceleration[n]
+        dt = trajectory_time[n + 1] - trajectory_time[n]  # dt
+        current_velocity = previous_velocity + trajectory_acceleration[n] * dt  # velocity
+        current_position = (
+            previous_position + (current_velocity + previous_velocity) / 2 * dt
+        )  # position
+
+        stream_segments.append(
+            StreamSegment(
+                current_position,
+                max([abs(current_velocity), abs(previous_velocity)]),
+                abs(current_accel),
+                dt,
+            )
+        )
+
+        # Record current values and previous positions for next step
+        previous_position = current_position
+        previous_velocity = current_velocity
+
+    return stream_segments
+
+
+class ZeroVibrationStreamGenerator:
+    """A class for creating stream motion with zero vibration input shaping theory."""
+
+    # pylint: disable=too-many-instance-attributes
+    # Ignore number of attributes warning because calling property setters in init is being counted
+    # as additional attributes
+
+    def __init__(
+        self,
+        resonant_frequency: float,
+        damping_ratio: float,
+        shaper_type: ShaperType = ShaperType.ZV,
+    ) -> None:
+        """
+        Initialize the class.
+
+        :param resonant_frequency: The target vibration resonant frequency in Hz.
+        :param damping_ratio: The target vibration damping ratio.
+        :param damping_ratio: Minimum timestep in pvt sequence in seconds.
+        """
+        self.resonant_frequency = resonant_frequency
+        self.damping_ratio = damping_ratio
+        self.shaper_type = shaper_type
+
+        # Initialize impulses as a single impulse with no delay for no shaping
+        self._impulse_times = [0.0]
+        self._impulses = [1.0]
+        self._impulses_updated = False
```

Comment:
Is this needed? Why don't you just update the impulses automatically every time the frequency or damping ratio gets set? The calculation doesn't look expensive and those parameters won't be updated much. There's a ton of if/else logic relating to this variable that could be removed too.

Reply:
I have opted to calculate them in get_impulses() and get_impulse_times(). You are right that it is not much computation so should be fine to recalculate every time. I'm not sure why I felt the need to do that in the first place.